### PR TITLE
feat(opd): support SGLang as teacher server for online policy distillation

### PR DESCRIPTION
**Pull request by nathon-lee**
Related issue: https://github.com/verl-project/verl/issues/5836

## What does this PR do?

Enables SGLang to be used as the **teacher inference server** in the Online
Policy Distillation (OPD) pipeline.  Previously only vLLM was supported as
teacher backend.

## Motivation

The verl 26Q2 roadmap lists "Support sglang/trt-llm teacher server" as a target
for the Online Policy Distillation track.  SGLang can offer better throughput
and lower latency than vLLM in certain hardware/model configurations, so making
the teacher backend pluggable increases flexibility.

### `verl/workers/rollout/sglang_rollout/async_sglang_server.py`

- **New `_extract_sglang_prompt_logprobs()`**: converts SGLang's
  `meta_info["input_token_logprobs"]` / `meta_info["input_top_logprobs"]`
  output into the `{"prompt_ids": ..., "prompt_logprobs": ...}` dict that
  `AsyncTeacherLLMServerManager.compute_teacher_logprobs_single()` expects
  (mirroring the shape conventions of vLLM's `extract_prompt_logprobs()`).

- **`SGLangHttpServer.generate()` now handles `prompt_logprobs`**: when this
  key is present in `sampling_params`, it is popped and mapped to SGLang's
  native `return_logprob=True` + `logprob_start_len=0` + `top_logprobs_num=K`
  before the request is forwarded to the engine.  The decoded results are
  attached to `extra_fields` via `_extract_sglang_prompt_logprobs()`.  The
  existing `logprobs` (output-token) path is unchanged.

### `verl/experimental/teacher_loop/teacher_manager.py`

- Removed the `raise NotImplementedError("vLLM does not support temperature for
  prompt_logprobs.")` guard from `_get_teacher_sampling_params()`.  The
  restriction was inaccurate (vLLM's `SamplingParams` accepts both fields) and
  blocked non-default temperature for all backends.  Updated the docstring to
  describe how `prompt_logprobs` is interpreted by each backend.

### `examples/on_policy_distillation_trainer/run_qwen_gsm8k_sglang_teacher.sh`

New example script: Qwen2.5-3B-Instruct as SGLang teacher, Qwen2.5-0.5B as
vLLM student, GSM8K dataset.  Mirrors the existing `run_qwen_gsm8k.sh` with
`distillation.teacher_model.inference.name=sglang`.

### CODE REVIEW

File: verl/workers/rollout/sglang_rollout/async_sglang_server.py [hunk 1]
```diff
@@ -57,6 +57,85 @@
...
+            ids = [int(token_id)]
+            lps = [float(logprob) if logprob is not None else 0.0]
+
+        prompt_ids_ls.append(ids)
+        prompt_logprobs_ls.append(lps)
+
+    # Append a dummy entry for the last prompt token (no logprob available).
+    k = max(num_prompt_logprobs, 1)
+    prompt_ids_ls.append([0] * k)
+    prompt_logprobs_ls.append([0.0] * k)
```
[the remark this answers is not in the document]
Comment (nathon-lee) [reply verl/workers/rollout/sglang_rollout/async_sglang_server.py]: Good catch, thanks. I updated _extract_sglang_prompt_logprobs to skip the first prompt token, keep row widths consistent, and append a single trailing dummy row so the output length matches len(sequence_ids). I also verified that SGLang’s input_top_logprobs is positionally aligned with input_token_logprobs.

File: examples/on_policy_distillation_trainer/run_qwen_gsm8k_sglang_teacher.sh [hunk 2]
```diff
@@ -0,0 +1,174 @@
+#!/usr/bin/env bash
```
Comment (wuxibin89) [inline examples/on_policy_distillation_trainer/run_qwen_gsm8k_sglang_teacher.sh]: Please reuse examples/on_policy_distillation_trainer/run_qwen_gsm8k.sh by change ROLLOUT_NAME.

**Comment by wuxibin89** [comment]
Please verify examples/on_policy_distillation_trainer/run_qwen_gsm8k.sh with sglang and report wandb metrics.
- DISTILLATION_LOSS_MODE=forward_kl_topk
- DISTILLATION_LOSS_MODE=k1

File: verl/experimental/teacher_loop/teacher_manager.py [hunk 3]
```diff
@@ -32,10 +32,18 @@ def _get_teacher_sampling_params(
     distillation_config: DistillationConfig,
     distillation_loss_config: DistillationLossConfig,
 ) -> dict[str, Any]:
-    """Get sampling parameters for teacher model when computing log probabilities for distillation."""
-    if distillation_config.teacher_model.inference.temperature != 1.0:
-        raise NotImplementedError("vLLM does not support temperature for prompt_logprobs.")
```
Comment (JacobHelwig) [inline verl/experimental/teacher_loop/teacher_manager.py:37]: The temperature in vLLM is only applied to logits of decoded tokens, not prompt tokens.  
1. https://github.com/vllm-project/vllm/pull/37082
2. https://discuss.vllm.ai/t/given-a-completion-text-for-a-fixed-prompt-text-how-to-calculate-the-log-prob-of-the-completion-text/2436

Comment (JacobHelwig) [reply verl/experimental/teacher_loop/teacher_manager.py:37]: How does sglang handle temperature for prompt logprobs?

File: verl/workers/rollout/sglang_rollout/async_sglang_server.py [hunk 4]
```diff
@@ -57,6 +57,56 @@
...
+    meta_info = output["meta_info"]
+
+    raw_token_logprobs = meta_info.get("input_token_logprobs") or []
+    raw_top_logprobs = meta_info.get("input_top_logprobs") or []
+
+    prompt_ids_ls: list[list[int]] = []
+    prompt_logprobs_ls: list[list[float]] = []
+
+    target_k = max(num_prompt_logprobs, 1)
+    use_top = num_prompt_logprobs > 0 and len(raw_top_logprobs) > 0
```
Comment (JacobHelwig) [inline verl/workers/rollout/sglang_rollout/async_sglang_server.py]: Will there be a correct case where num_prompt_logprobs > 0 but len(raw_top_logprobs) == 0? If not, then maybe should raise an error

File: verl/workers/rollout/sglang_rollout/async_sglang_server.py [hunk 5]
```diff
@@ -57,6 +57,56 @@
...
+    for i in range(1, len(raw_token_logprobs)):
+        logprob, token_id, _ = raw_token_logprobs[i]
+
+        if use_top:
+            top_entries = raw_top_logprobs[i] if i < len(raw_top_logprobs) else []
+            ids = [int(tid) for _, tid, _ in top_entries[:num_prompt_logprobs]]
+            lps = [float(lp) for lp, _, _ in top_entries[:num_prompt_logprobs]]
+
+            if len(ids) == 0:
+                ids = [int(token_id)]
```
Comment (JacobHelwig) [inline verl/workers/rollout/sglang_rollout/async_sglang_server.py]: Will there be a correct case where we can enter this block?

File: verl/workers/rollout/sglang_rollout/async_sglang_server.py [hunk 6]
```diff
@@ -57,6 +57,56 @@
...
+                lps = [float(logprob) if logprob is not None else 0.0]
+        else:
+            ids = [int(token_id)]
+            lps = [float(logprob) if logprob is not None else 0.0]
+
+        while len(ids) < target_k:
+            ids.append(0)
```
Comment (JacobHelwig) [inline verl/workers/rollout/sglang_rollout/async_sglang_server.py]: Will there be a correct case where we can enter this block?

File: verl/workers/rollout/sglang_rollout/async_sglang_server.py [hunk 7]
```diff
@@ -393,16 +443,32 @@ async def generate(
         sampling_params["max_new_tokens"] = max_new_tokens
         return_logprob = sampling_params.pop("logprobs", False)
 
+        # prompt_logprobs: number of top-K token logprobs to return for each input token position.
+        # This is used by the OPD (Online Policy Distillation) teacher server to provide
```
Comment (JacobHelwig) [inline verl/workers/rollout/sglang_rollout/async_sglang_server.py]:
```suggestion
        # This is used by the OPD (On-Policy Distillation) teacher server to provide
```

**Review by JacobHelwig — CHANGES_REQUESTED** [review]

File: verl/workers/rollout/sglang_rollout/async_sglang_server.py (hunk 5 above)
Comment (nathon-lee) [reply verl/workers/rollout/sglang_rollout/async_sglang_server.py]: "It is possible to enter (top_entries may be empty/missing), but top-k distillation cannot be silent fallback, so I changed it to directly report an error when encountering empty top_entries."

File: verl/workers/rollout/sglang_rollout/async_sglang_server.py (hunk 6 above)
Comment (nathon-lee) [reply verl/workers/rollout/sglang_rollout/async_sglang_server.py]: The padding branch will be entered (the number of top_entries may be less than K). This is to ensure that the tensor width is consistent at each position and to avoid ragged tensors.

File: examples/on_policy_distillation_trainer/run_qwen_gsm8k_sglang_teacher.sh (hunk 2 above)
Comment (nathon-lee) [reply examples/on_policy_distillation_trainer/run_qwen_gsm8k_sglang_teacher.sh]: ok

**Comment by wuxibin89** [comment]
https://github.com/volcengine/verl/issues/6120
